core/src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -682,14 +700,226 @@ decodeUtf8Lax = decodeUtf8
 -- Decoding Utf16
 -------------------------------------------------------------------------------
 
--- | Decode a UTF-16 little endian encoded bytestream to a stream of Unicode
--- characters. The function throws an error if an invalid codepoint is
--- encountered.
+data MkEvenW8ChunksState s w8 arr
+    = MECSInit s
+    | MECSBuffer w8 s
+    | MECSYieldAndInit arr s
+    | MECSYieldAndBuffer arr w8 s
+
+{-# INLINE_NORMAL mkEvenW8Chunks #-}
+mkEvenW8Chunks :: Monad m => Stream m (Array Word8) -> Stream m (Array Word8)
+mkEvenW8Chunks (D.Stream step state) = D.Stream step1 (MECSInit state)
+
+    where
+
+    {-# INLINE_LATE step1 #-}
+    step1 gst (MECSInit st) = do
+        r <- step (adaptState gst) st
+        return $
+            case r of
+                Yield arr st1 | Array.length arr == 0 -> Skip (MECSInit st1)
+                Yield arr st1 | Array.length arr == 1 ->
+                    let remElem = Array.getIndexUnsafe 0 arr
+                     in Skip (MECSBuffer remElem st1)
+                Yield arr st1 ->
+                    let len = Array.length arr
+                     in if (len .&. 1) == 1
+                        then let arr1 = Array.getSliceUnsafe 0 (len - 1) arr
+                                 remElem = Array.getIndexUnsafe (len - 1) arr
+                              in Yield arr1 (MECSBuffer remElem st1)
+                        else Yield arr (MECSInit st1)
+                Skip s -> Skip (MECSInit s)
+                Stop -> Stop
+    step1 gst (MECSBuffer remElem st) = do
+        r <- step (adaptState gst) st
+        return $
+            case r of
+                Yield arr st1 | Array.length arr == 0 ->
+                                  Skip (MECSBuffer remElem st1)
+                Yield arr st1 | Array.length arr == 1 ->
+                    let fstElem = Array.getIndexUnsafe 0 arr
+                        w16 = Array.fromList [remElem, fstElem]
+                     in Yield w16 (MECSInit st1)
+                Yield arr st1 ->
+                    let len = Array.length arr
+                     in if (len .&. 1) == 1
+                        then let arr1 = Array.getSliceUnsafe 1 (len - 1) arr
+                                 fstElem = Array.getIndexUnsafe 0 arr
+                                 w16 = Array.fromList [remElem, fstElem]
+                              in Yield w16 (MECSYieldAndInit arr1 st1)
+                        else let arr1 = Array.getSliceUnsafe 1 (len - 2) arr
+                                 fstElem = Array.getIndexUnsafe 0 arr
+                                 lstElem = Array.getIndexUnsafe (len - 1) arr
+                                 w16 = Array.fromList [remElem, fstElem]
+                              in Yield w16
+                                     (MECSYieldAndBuffer arr1 lstElem st1)
+                Skip s -> Skip (MECSBuffer remElem s)
+                Stop -> Stop -- Here the last Word8 is lost
```

Comment:
This has nothing to do with utf16 decoding.

Reply:
This is only used in the benchmark for now anyway, so I don't want to spend more time on this. We can change this if this use-case ever arises.

Reply:
If it has nothing to do with it then we should not use it in decoding :-) Because decoding should fail if it finds the input is truncated. If this drops silently, decoder won't even know that some part of the input was dropped, it will just happily use whatever it got.